share/eds/cache/accessor_cache.go

```diff
@@ -0,0 +1,189 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"reflect"
+	"sync"
+
+	"github.com/filecoin-project/dagstore"
+	"github.com/filecoin-project/dagstore/shard"
+	lru "github.com/hashicorp/golang-lru"
+)
+
+var _ Cache = (*AccessorCache)(nil)
+
+// AccessorCache implements the Cache interface using an LRU cache backend.
+type AccessorCache struct {
+	// The name is a prefix that will be used for cache metrics if they are enabled.
+	name string
+	// stripedLocks prevents simultaneous RW access to the blockstore cache for a shard. Instead
+	// of using only one lock or one lock per key, we stripe the shard keys across 256 locks. 256 is
+	// chosen because it 0-255 is the range of values we get looking at the last byte of the key.
+	stripedLocks [256]sync.Mutex
```

Comment:
Stripped locks are really effective, except that an array of mutexes as is will cause what's called false sharing.

Although multiple routines will get to work concurrently on different shards and not block each other (_except for same shard routines_), some "unrelated" routines will "block" (_hardware level_) each other due to CPU cacheline, basically because they are -_say_- trying to access two different mutexes each, which live on the same cacheline. Although they are not sharing any resources, the way how the cpu cacheline works will cause them to "falsely share" that cache line.

Since an array is contiguous space in memory, and a mutex is exactly 8 bytes, 8 of them will live in the same cacheline, so with this specific case, if only 64 shards are being accessed by 64 routines, 1 routine for each shard, and let's say half of the routines want to access the shards from 0-31 and are on cpu core 1 and the other half that is trying to access 31-63 and are on cpu core 2, although in theory each should get their respective lock fine, what will happen underneath is that whenever the first half of the routines will try to get a lock from the array, the second half is basically blocked waiting for core 1 to send back the newly modified cacheline, and so on back end forth between the cores for each lock access and modification.

If we crank it up to 255 and dunno how many routines, the performance will see an impact.

So to avoid this, each lock should be on its own cache line, which is possible in go by using `cpu.CacheLinePad`, but then it would require the mutex to be embedded in a new type and also embeds `cpu.CacheLinePad`:

```go
type paddedLock struct {
     sync.Mutex
     _ cpu.CacheLinePad
}

// ...

var stripedLocks [256]paddedLock
```

Correct me on this

Reply:
@derrandz, I think you are correct with the false sharing argument. We use stripped locking in several places in the code, so it would be nice to make an issue about making all of them cache padded and then PR which I'll happily review.

Reply:
🫡